### PR TITLE
Deletion Marker Fixup

### PR DESCRIPTION
**Pull request by mridul-ms**
## Why this change is required?
This change is required to support Deletion Markers in MSIX similar to App-V as of today. 

## What changed?
1. RegLegacyFixups : 
    a. Initialize Fixups - changes to store deletion marker fixup from config.
    b. RegistryFixups - detour for APIs
2. Tests

##How this change has been tested?
Added automated tests for each detoured API.

### CODE REVIEW

File: fixups/RegLegacyFixups/RegistryFixups.cpp [hunk 1]
```diff
@@ -518,6 +724,622 @@ DECLARE_STRING_FIXUP(RegOpenKeyTransactedImpl, RegOpenKeyTransactedFixup);
 
 
 
+/// <summary>
+/// Method to interpret Regitry Path
+/// </summary>
+/// <param name="regPath"></param>
+/// <returns></returns>
+std::string InterpretRegistryPath(std::string regPath)
```
Comment (ravishroshanms) [inline fixups/RegLegacyFixups/RegistryFixups.cpp:732]: Why have created this new function and not used ReplaceRegistrySyntax?

File: tests/scenarios/RegLegacyTest/config.json [hunk 2]
```diff
@@ -21,6 +21,23 @@
...
+                  "type": "DeletionMarker",
+                  "hive": "HKLM",
+                  "key": "^SOFTWARE.*",
+                  "values": [ "DeletionMarker" ]
+                },
+                {
+                  "type": "DeletionMarker",
+                  "hive": "HKCU",
+                  "key": "^Software.*",
+                  "values": [ "DeletionMarker" ]
```
Comment (ravishroshanms) [inline tests/scenarios/RegLegacyTest/config.json:39]: Can you add a test where Values is empty?

**Review by ravishroshanms — CHANGES_REQUESTED** [review]

File: tests/scenarios/RegLegacyTest/config.json (hunk 2 above)
Comment (mridul-ms) [reply tests/scenarios/RegLegacyTest/config.json:39]: Sure, I'll add a configuration with empty values, and use existing tests to cover. it.

File: fixups/RegLegacyFixups/RegistryFixups.cpp [hunk 3]
```diff
@@ -320,9 +328,175 @@ bool RegFixupFakeDelete(std::string keypath)
                             Log("[%d] Bad Regex pattern ignored in RegLegacyFixups.\n", RegLocalInstance);
 #endif
                         }
+                    }
+                }
+                break;
+            }
```
[the remark this answers is not in the document]
Comment (mridul-ms) [reply fixups/RegLegacyFixups/RegistryFixups.cpp:334]: No change has been made here. Diff is shown because same brackets structure is there in below method. Git shows is this way :)

**Comment by dhoehna** [comment]
What is a deletion marker?

File: fixups/RegLegacyFixups/InitializeFixup.cpp [hunk 4]
```diff
@@ -256,6 +255,63 @@ void InitializeConfiguration()
 
                             specItem.remediationRecords.push_back(recordItem);
                         }
+
+                        // Look for DeletionMarker remediation
+                        else if (type.compare(L"DeletionMarker") == 0)
+                        {
+                            Log("RegLegacyFixups:      is DeletionMarker\n");
```
Comment (dhoehna) [inline fixups/RegLegacyFixups/InitializeFixup.cpp:262]: what is with all the spaces?  Are you trying to do a tab?

File: fixups/RegLegacyFixups/InitializeFixup.cpp [hunk 5]
```diff
@@ -256,6 +255,63 @@ void InitializeConfiguration()
...
+                            recordItem.remeditaionType = Reg_Remediation_type_DeletionMarker;
+
+                            // Look for hive - data
+                            auto hiveType = regItemObject.try_get("hive")->as_string().wstring();
```
Comment (dhoehna) [inline fixups/RegLegacyFixups/InitializeFixup.cpp:266]: try_get returns nullptr if the item can't be found.  This will crash.

File: fixups/RegLegacyFixups/InitializeFixup.cpp [hunk 6]
```diff
@@ -256,6 +255,63 @@ void InitializeConfiguration()
...
+                            traceDataStream << " hive: " << hiveType << " ;";
+                            Log(L"Hive:      %Ls\n", hiveType.data());
+
+                            if (hiveType.compare(L"HKCU") == 0)
```
Comment (dhoehna) [inline fixups/RegLegacyFixups/InitializeFixup.cpp:270]: what happens in hivetype is "hkcu"?  Looks like  I can't tell if .compare is case sensitive or not.

File: fixups/RegLegacyFixups/RegistryFixups.cpp (hunk 1 above)
Comment (ravishroshanms) [reply fixups/RegLegacyFixups/RegistryFixups.cpp:732]: I don't see that you are doing anything different than the purpose of this method. Could you please converge your separate thread and make sure we don't have duplication of code. Thanks

File: fixups/RegLegacyFixups/InitializeFixup.cpp (hunk 6 above)
Comment (ravishroshanms) [reply fixups/RegLegacyFixups/InitializeFixup.cpp:270]: There is a try catch block from where we are trying to attach the dll to the process(In DllMain of PsfRuntime). These are all mandatory fields, and we don't want to load the fixup if anything fails here.

File: fixups/RegLegacyFixups/InitializeFixup.cpp (hunk 5 above)
Comment (ravishroshanms) [reply fixups/RegLegacyFixups/InitializeFixup.cpp:266]: There is a try catch block from where we are trying to attach the dll to the process(In DllMain of PsfRuntime). These are all mandatory fields, and we don't want to load the fixup if anything fails here.

File: fixups/RegLegacyFixups/InitializeFixup.cpp (hunk 6 above)
Comment (mridul-ms) [reply fixups/RegLegacyFixups/InitializeFixup.cpp:270]: Yes, this is not case sensitive. This shall be covered in public documentation.

File: fixups/RegLegacyFixups/InitializeFixup.cpp (hunk 5 above)
Comment (mridul-ms) [reply fixups/RegLegacyFixups/InitializeFixup.cpp:266]: Correct, these are all mandatory field. This will be also covered in public documentation.

File: fixups/RegLegacyFixups/InitializeFixup.cpp (hunk 4 above)
Comment (mridul-ms) [reply fixups/RegLegacyFixups/InitializeFixup.cpp:262]: Yes.

**Comment by mridul-ms** [comment]
> What is a deletion marker?

@dhoehna please refer this blog to understand on deletion markers.
https://www.tmurgent.com/TmBlog/?p=1188

File: fixups/RegLegacyFixups/RegistryFixups.cpp (hunk 1 above)
Comment (mridul-ms) [reply fixups/RegLegacyFixups/RegistryFixups.cpp:732]: As discussed with Rahul, we need to understand more on existing method (maybe existing things are done for a purpose). I have created an issue to track this. 
https://github.com/microsoft/MSIX-PackageSupportFramework/issues/246
Until the issue is closed, I'm not looking to converge these methods.

File: fixups/RegLegacyFixups/InitializeFixup.cpp (hunk 4 above)
Comment (dhoehna) [reply fixups/RegLegacyFixups/InitializeFixup.cpp:262]: Please use `\t`

File: fixups/RegLegacyFixups/InitializeFixup.cpp (hunk 6 above)
Comment (dhoehna) [reply fixups/RegLegacyFixups/InitializeFixup.cpp:270]: Please change this to a case-insensitive compare.  If `hiveType` is `hkcu` this should still work.

File: fixups/RegLegacyFixups/RegistryFixups.cpp [hunk 7]
```diff
@@ -320,9 +328,175 @@ bool RegFixupFakeDelete(std::string keypath)
...
+                        catch (...)
+                        {
+                            psf::TraceLogExceptions("RegLegacyFixupException", "Bad Regex pattern ignored in RegLegacyFixups. Hive: HKCU");
+#ifdef _DEBUG
+                            Log("[%d] Bad Regex pattern ignored in RegLegacyFixups.\n", RegLocalInstance);
+#endif
+                        }
+                    }
+                    break;
+                case Modify_Key_Hive_Type_HKLM:
```
[the remark this answers is not in the document]
Comment (mridul-ms) [reply fixups/RegLegacyFixups/RegistryFixups.cpp:439]: Sure, added

File: fixups/RegLegacyFixups/InitializeFixup.cpp (hunk 4 above)
Comment (mridul-ms) [reply fixups/RegLegacyFixups/InitializeFixup.cpp:262]: Sure, changed.

File: tests/scenarios/RegLegacyTest/config.json (hunk 2 above)
Comment (mridul-ms) [reply tests/scenarios/RegLegacyTest/config.json:45]: Added.

**Comment by mridul-ms** [comment]
@dhoehna In App-V, a deletion marker for registry is a special marker that is created (during monitoring of changes during creation of App-V package with App-V sequencer), in the registries (with a special ‘Type’) when a registry key is deleted from a virtual environment. This marker is stored in the virtual registry and is used by the App-V client to ensure that the deleted registry key is not included in the virtual environment when the package is streamed or deployed to client computers.

**Comment by rahul89dadhich** [comment]
[like]  Rahul Dadhich reacted to your message:
________________________________

**Review by ravishroshanms — APPROVED** [review]

**Review by kyelamarthi1 — APPROVED** [review]
LGTM

File: fixups/RegLegacyFixups/InitializeFixup.cpp (hunk 6 above)
Comment (mridul-ms) [reply fixups/RegLegacyFixups/InitializeFixup.cpp:270]: Hi Darren, currently this is same for all other configs. We can take this as a separate task to have case insensitive comparisons for such strings. For now, keeping it as is.

**Comment by dhoehna** [comment]
Mridul and I discussed offline about case-insensitive compare.

@mridul-ms Please make a bug on PSF for changing these comparisons to case-insensitive.

File: fixups/RegLegacyFixups/RegistryFixups.cpp [hunk 8]
```diff
@@ -45,10 +52,10 @@ std::string ReplaceRegistrySyntax(std::string regPath)
     {
         if (regPath.length() > 18)
         {
-            size_t offsetAfterSid = regPath.find('\\', 19);
+            size_t offsetAfterSid = regPath.find('\\', 18);
```
Comment (npuvvada) [inline fixups/RegLegacyFixups/RegistryFixups.cpp:55]: Wouldn't this find offsetBeforeSid instead of offsetAfterSid

**Comment by npuvvada** [comment]
Please add relevant documentation in RegLegacy readme file to explain purpose of this fixup and how it can be configured.

File: fixups/RegLegacyFixups/RegistryFixups.cpp (hunk 8 above)
Comment (mridul-ms) [reply fixups/RegLegacyFixups/RegistryFixups.cpp:55]: Do we get Sid in HKLM path?

**Comment by mridul-ms** [comment]
@dhoehna Created an issue for Case insensitive comparision.
https://github.com/microsoft/MSIX-PackageSupportFramework/issues/247

File: fixups/RegLegacyFixups/RegistryFixups.cpp (hunk 8 above)
Comment (npuvvada) [reply fixups/RegLegacyFixups/RegistryFixups.cpp:55]: Agreed, it looks \Registry\Machine entries do not have SID which is contradictory from the variable name 'offsetAfterSid'.
